### ISSUE TITLE
[Question] - In memory files

### ISSUE THREAD

**Issue by dazinator** [pahen/madge#186]
Is it possible to provide the files to `madge` from an in-memory store rather than it going off to disk?

**Comment by mrjoelkemp** [comment]
Hey! Interesting question. You might be able to use mock-fs and invoke madge programmatically. As in, invoke mockfs then use madge in a node script. I'm not sure how you'd do it via a cli invocation of madge.

**Comment by dazinator** [comment]
Thanks I will take a look at `mock-fs`.
Other libraries I have integrated with so far, like `rollup` and` requirejs` have plugins for this use case. For example: https://github.com/rxaviers/requirejs-memfiles or https://www.npmjs.com/package/rollup-plugin-hypothetical - however if this mock-fs will do the job without any special effort on the part of `madge` then so much the better :-)

**Comment by dazinator** [comment]
Actually, perhaps there is something that lets me create a virtual filesystem, with in memory files layered on top of physical files. Let me try: https://www.npmjs.com/package/memfs

**Comment by dazinator** [comment]
Nope, so I managed to do some interesting stuff with https://github.com/streamich/unionfs like create a virtual filesystem (object implementing fs API) that is backed by multiple sources such as mem-fs as well as node's physical fs. However I have no way to pass this union-fs instance to `madge`. The `mock-fs` shadows nodes own `fs` implementation meaning there is no need to pass an instance to `madge` - but that has the side effect that all dynamic imports then fail, as `node_module` dependencies can't be loaded, and this is just too much to mock up just to be able to provide a couple of module file contents from memory.

I think at this point, the nicest change would be to allow an `fs` instance to be provided via `options` and if it is, then `madge` can use that for file system access rather then node's default `fs`. This will provide great flexibility by allowing `union-fs` instances to be passed in that blend files from various sources if need be.. What do you think?

**Comment by mrjoelkemp** [comment]
Thanks for trying a few solutions. What's the reason for needing the in memory version of the files? Can you provide a bit more context that led to the need?

Curious if it's super custom or a generic use case that we're just encountering. Thanks in advance.

**Comment by mrjoelkemp** [comment]
Thanks a ton for the context. That's helpful info to reason about the need.

I think support is required at many layers for this. Madge may have some fs interactions, but the core of it is likely done at deeper layers like node-dependency-tree, node-precinct, and node-filing-cabinet. I'd start reading through dependency-tree and you'll probably need to pass an fs object as an option to the call to precinct.paperwork(). Then submit a patch to node-precinct to use that fs object instead of the default fs import. That may be all that you need in core changes; changes to madge and dependency-tree would be accepting and propagating the fs option down. 

Happy to review PRs for dependency-tree all the way down. You'll need @pahen's blessing for the madge proper changes. 

As a last part towards your previous workarounds, you can try using https://www.npmjs.com/package/rewire to theoretically replace the fs used all the way down. Maybe worth a try before you have to go deep.

**Comment by dazinator** [comment]
Thanks a lot for the `Rewire` suggestion, if I can get that working it shpulf be a much easier path, I'll give that a go first ;-)

**Comment by pahen** [comment]
Interesting use case for Madge. I'm gladly accepting a PR to support this option :)

**Comment by dazinator** [comment]
@mrjoelkemp `rewiring` the various modules is a bit tricky. Some of the `require(fs)` calls are themselves in functions, making them hard to replace with `rewire` (based on my novice skill level anyway!)

So I am starting my journey through the stack like you mentioned, have started with [precinct paperwork ](https://github.com/dependents/node-precinct/pull/52) :-)

**Comment by pahen** [comment]
Is this something you still work on @dazinator ? Or can we close this issue?

**Comment by dazinator** [comment]
Still going.. progress stopped for a while whilst waiting on downstream dependency but that's unblocked now so I intend to resume this in the near future.

**Comment by pahen** [comment]
Ok, cool :)

**Comment by dazinator** [comment]
Hi @iddan - no I am no longer actively working on this.
I got quite far with it, including getting a bunch of the underlying dependencies changed to support the feature and then building upon them, with tests.

I was still working in completing `node-filing-cabinet` however, as per this issue here - much of it was working but there were some remaining tasks: https://github.com/dependents/node-filing-cabinet/pull/68

Also the `file-exists` dependency the author never replied. I forked it and made changes to allow a custom `fs` to be supplied, published my fork to npm and built on that version: https://github.com/dazinator/file-exists

For me, the original use case is not pressing, - would have been a nice to have, but I have now moved on. If someone did want to pick this up, I could provide info, otherwise, it could be dropped.
